tests/contrib/flask/test_flask_snapshot.py

```diff
@@ -0,0 +1,104 @@
+import os
+import signal
+import subprocess
+import time
+from typing import Dict
+from typing import Generator
+from typing import List
+
+import pytest
+import tenacity
+
+from tests.webclient import Client
+
+
+@pytest.fixture
+def flask_port():
+    # type: () -> str
+    return "8000"
+
+
+@pytest.fixture
+def flask_wsgi_application():
+    # type: () -> str
+    return "tests.contrib.flask.app:app"
+
+
+@pytest.fixture
+def flask_command(flask_wsgi_application, flask_port):
+    # type: (str, str) -> List[str]
+    cmd = (
+        "uwsgi --enable-threads --lazy-apps "
+        "--import=ddtrace.bootstrap.sitecustomize "
+        "--master --processes=1 --http 0.0.0.0:%s "
+        "--module %s"
+    ) % (flask_port, flask_wsgi_application)
+    return cmd.split(" ")
+
+
+@pytest.fixture
+def flask_env():
+    # type: () -> Dict[str, str]
+    env = os.environ.copy()
+    env.update(
+        {
+            # Avoid noisy database spans being output on app startup/teardown.
+            "DD_TRACE_SQLITE3_ENABLED": "0",
+        }
+    )
+    return env
+
+
+@pytest.fixture
+def flask_client(flask_command, flask_env, flask_port):
+    # type: (str, Dict[str, str], str) -> Generator[Client, None, None]
+    # Copy the env to get the correct PYTHONPATH and such
+    # from the virtualenv.
+    # webservers might exec or fork into another process, so we need to os.setsid() to create a process group
+    # (all of which will listen to signals sent to the parent) so that we can kill the whole application.
+    proc = subprocess.Popen(
+        flask_command,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+        close_fds=True,
+        env=flask_env,
+        preexec_fn=os.setsid,
+    )
+    try:
```

Comment:
Just trying to understand, why for other integration snapshot tests do we not have to worry about servers not being up yet? In addition, why do we need to wait for the test agent here because of the race condition? I haven't seen that in other integrations. Is it because we're testing streams here rather than just a normal request?

Reply:
> why for other integration snapshot tests do we not have to worry about servers not being up yet?

We should! This is probably the result of some other snapshot tests being flaky, because servers aren't ready yet.

> why do we need to wait for the test agent here because of the race condition?

There is a slight race condition between when the traces are sent from the application and when we ask the test agent for the traces (the test agent might be processing the payload when we ask).

> I haven't seen that in other integrations. Is it because we're testing streams here rather than just a normal request?

We should be doing the same in other integrations, too since the race exists. We just haven't gotten around to it yet. These are some new ideas introduced in #3705 that haven't made it in quite yet. It doesn't have anything to do with streaming in particular.